draft-ietf-webpush-encryption.md

```diff
@@ -108,170 +101,277 @@ encryption scheme by applications that use push messaging.  An agent that only
-the URI of the subscription.  The private key MUST remain secret.
 
-This key pair is used with the Diffie-Hellman key exchange as described in
-Section 4.2 of [I-D.ietf-httpbis-encryption-encoding].
+## Key and Secret Distribution
 
-A User Agent MUST generate and provide a public key for the scheme described in
-{{mti}}.
+The application using the subscription distributes the subscription public key
+and authentication secret to an authorized Application Server.  This could be
+sent along with other subscription information that is provided by the User
+Agent, such as the push subscription URI.
 
-The public key MUST be accompanied by a key identifier that can be used in the
-"keyid" parameter to identify which key is in use.  Key identifiers need only be
-unique within the context of a subscription.
+An application MUST use an authenticated, confidentiality protected
+communications medium for this purpose.  In addition to the reasons described in
+{{!I-D.ietf-webpush-protocol}}, this ensures that the authentication secret is
+not revealed to unauthorized entities, which can be used to generate push
+messages that will be accepted by the User Agent.
 
+Most applications that use push messaging have a pre-existing relationship with
+an Application Server.  Any existing communication mechanism that is
+authenticated and provides confidentiality and integrity, such as HTTPS
+{{?RFC2818}}, is sufficient.
 
-## Diffie-Hellman Group Information
 
-As described in [I-D.ietf-httpbis-encryption-encoding], use of Diffie-Hellman
-for key agreement requires that the receiver provide clear information about its
-chosen group and the format for the "dh" parameter with each potential sender.
+# Push Message Encryption {#encryption}
 
-This document only describes a single ECDH group and point format, described in
-{{mti}}.  A specification that defines alternative groups or formats MUST
-provide a means of indicating precisely which group and format is in use for
-every public key that is provided.
+Push message encryption happens in four phases:
 
+* The input keying material used for deriving the content encryption keys used
+  for the push message is derived using elliptic-curve Diffie-Hellman {{ECDH}}
+  ({{dh}}).
 
-## Key Distribution
+* This is then combined with the application secret to produce the input keying
+  material used in {{!I-D.ietf-httpbis-encryption-encoding}} ({{combine}}).
+
+* A content encryption key and nonce are derived using the process in
+  {{!I-D.ietf-httpbis-encryption-encoding}} with an expanded context string
+  ({{context}}).
+
+* Encryption or decryption follows according to
+  {{!I-D.ietf-httpbis-encryption-encoding}}.
+
+The key derivation process is summarized in {{summary}}.  Restrictions on the
+use of the encrypted content coding are described in {{restrict}}.
+
+
+## Diffie-Hellman Key Agreement {#dh}
+
+For each new subscription that the User Agent generates for an application, it
```

Comment:
I almost went through and removed Title Case Names throughout, but I guess now that I decided to keep them...